Adding unit tests to cover the project-to-project reference matrix wi…

johnbeisner opened the pull request:
Adding unit tests to cover the project-to-project reference matrix with .NET Standard/.NET Framework/.NET Core 1.x/.NET Core 2.0

Removed duplicate unit tests from 'GivenThatWeWhatToReferenceAProject' that are now covered in 'GivenThatWeWhatToVerifyProjectReferenceCompat'

https://github.com/dotnet/sdk/issues/973 [issue contains the test matrix]

Commit:
Adding unit tests to cover the project-to-project reference matrix with .NET Standard/.NET Framework/.NET Core 1.x/.NET Core 2.0

Removed duplicate unit tests from 'GivenThatWeWhatToReferenceAProject' that are now covered in 'GivenThatWeWhatToVerifyProjectReferenceCompat'

https://github.com/dotnet/sdk/issues/973

dnfclas commented:
@__johnbeisner__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

eerhardt commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,213 @@
+                //  Workaround for .NET Core 2.0
+                if (target.Equals("netcoreapp2.0", StringComparison.OrdinalIgnoreCase) || target.Equals("net462", StringComparison.OrdinalIgnoreCase))
+                {
+                    ret.RuntimeFrameworkVersion = "2.0.0-beta-001689-00";
```
Let's use `RepoInfo.NetCoreApp20Version` instead of hard-coding this.  That way it gets updated periodically.

johnbeisner replied:
updated...

eerhardt commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,213 @@
+            {
+                ret.TargetFrameworks = target;
+                //  Workaround for .NET Core 2.0
+                if (target.Equals("netcoreapp2.0", StringComparison.OrdinalIgnoreCase) || target.Equals("net462", StringComparison.OrdinalIgnoreCase))
```
Why `net462`?

johnbeisner replied:
net462 is part of the matrix; please refer to https://github.com/dotnet/sdk/issues/973

eerhardt replied:
I was a little too terse above.

The comment says `//  Workaround for .NET Core 2.0`, and we are setting `ret.RuntimeFrameworkVersion = "2.0.0-beta-001689-00";`, which should have no impact on `net462`.  So why are we checking for `net462` here?

johnbeisner replied:
OIC; I will update the comment - "2.0.0-beta-001689-00" does indeed have an impact on net462; I think we should not need to set the RuntimeframeworkVersion for any of these.

nguerrera replied:
It should not impact net462. What is the symptom when removed from this condition?

eerhardt replied:
Can you explain `"2.0.0-beta-001689-00" does indeed have an impact on net462;`?  I'm not making the connection on how this property will impact net462.

eerhardt commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,213 @@
+        [InlineData("netcoreapp2.0", "netstandard2.0", true)]
+        [InlineData("netcoreapp2.0", "netcoreapp1.0", true)]
+        [InlineData("netcoreapp2.0", "netcoreapp1.1", true)]
+        [InlineData("netcoreapp2.0", "netcoreapp2.0", true)]
```
How long do these tests take to run?  Do we think the value we are getting out of every single permuation is worth the length of test time?

eerhardt replied:
Doing some looking at the Jenkins runs, it looks like these tests are doubling the job time in Jenkins.

Runs without your change take between 7-8 minutes:
https://ci.dot.net/job/dotnet_sdk/job/master/job/release_windows_nt_prtest/217/
https://ci.dot.net/job/dotnet_sdk/job/master/job/release_windows_nt_prtest/218/

Your runs are taking 16 minutes:
https://ci.dot.net/job/dotnet_sdk/job/master/job/release_windows_nt_prtest/223/

Are we getting that much value out of these tests?

livarcocc replied:
These tests cover the whole supported reference matrix that we have. They are our guarantee that we are not regressing on that promise. I think that is a very good easy of mind thing over machine time. Also, it adds good coverage regarding backwards compatibility with old 1.0 TFMs, which is something else we really want.

livarcocc replied:
We chatted about some strategies to try to reduce the running time of these tests during stand-up. Let's see how much improvement we can get out of them.

johnbeisner replied:
I agree that we could be more efficient with these added unit tests; I am looking into combining similar references sets into a single mutli-targeted unit test.

johnbeisner replied:
Unofficial time-to-complete deltas:
debug_windows_nt_fullframework_prtest   +14 seconds
debug_windows_nt_prtest                 +1 minute 50 seconds
release_windows_nt_fullframework_prtest -28 seconds
release_windows_nt_prtest               +1 minute 14 seconds

eerhardt commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,213 @@
+        [InlineData("netcoreapp2.0", "netcoreapp1.1", true)]
+        [InlineData("netcoreapp2.0", "netcoreapp2.0", true)]
+
+        public void It_checks_for_valid_project_reference_compat(string referencerTarget, string dependencyTarget, bool buildSucceeds)
```
I only ever see `true` being passed in to `buildSucceeds`. Do we need this parameter?

johnbeisner replied:
Yes, the next work item is to test the negative cases.

Commit:
Removing the platform filter; using "RepoInfo.NetCoreApp20Version" instead of explictly setting the value.

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,204 @@
+            {
+                ret.TargetFrameworks = target;
+                //  Workaround for .NET Core 2.0 and .NET Framework 4.6.2
+                if (target.Equals("netcoreapp2.0", StringComparison.OrdinalIgnoreCase) || target.Equals("net462", StringComparison.OrdinalIgnoreCase))
```
We're having trouble communicating on this. The RuntimeFrameworkVersion property is expected to have zero impact on net462 projects. The condition and comment should be limited to netcoreapp2.

johnbeisner replied:
'The RuntimeFrameworkVersion property is expected to have zero impact on net462 projects' - the observed testing behaviors are in direct opposition to this statement - without the entry all SDK "net462" & classical .NET Framework "v4.6.2" unit tests fail; with the entry all unit test pass. The qualifier is that initial testing was done on a VM I created. The created VM may or may not be representative of the Jenkins testing bed image.

johnbeisner replied:
On Jenkins the 'RuntimeFrameworkVersion' property is not necessary for net462; removed for net462

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToReferenceAProject.cs:
```diff
@@ -41,15 +41,13 @@ public enum ReferenceBuildResult
         [InlineData("netcoreapp1.1;net45", true, "netstandard1.4", true, false, false)]
         [InlineData("netcoreapp1.1;net46", true, "net45;netstandard1.6", true, true, true)]
         [InlineData("netcoreapp1.1;net45", true, "net46;netstandard1.6", true, false, false)]
-        [InlineData("v4.6.1", false, "netstandard1.4", true, true, true)]
```
Why are two tests here removed? I see there's overlap between this and new tests but don't understand what makes these two and only these two redundant.

johnbeisner replied:
The unit tests in 'GivenThatWeWantToReferenceAProject' can be broken down into 4 categories:
1. negative single-target unit tests
2. negative multi-target unit tests
3. positive single-target unit tests
4. positive multi-target unit tests

The removed two are in category 3. 

Although the 'v4.6.1/netstandard1.4' is not 'exactly' redundant as the new unit test uses the SDK csproj construction where the removed unit test used the classical method.

Commit:
Currently can't build projects targeting .NET Framework on non-Windows: https://github.com/dotnet/sdk/issues/335

Commit:
Trying Jenkins PR runs with no RuntimeFrameworkVersion defined for net4.6.2

Commit:
GivenThatWeWantToVerifyProjectReferenceCompat.cs:
-Combine the 104 compat positive test matrix into 6 unit tests.
-Using a 'TestID' to avoid breaking the pathlength constraint.
TestProject.cs:
-Referencing projects that target a library were not exorcising P2P project references; fixed.

johnbeisner commented:
@nguerrera, @eerhardt, @dsplaisted please take an second look.

The combining of the unit tests into 'buckets' worked well. 104 tests => 16 tests; an increase of ~8 mins => ~1.5 minutes [on the longest queue]. Also fixed a test construction oversight; see: TestProject.cs

eerhardt commented on test/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:
```diff
@@ -239,8 +239,15 @@ namespace {this.Name}
+";
+                    foreach (var dependency in this.ReferencedProjects)
+                    {
+                        source += $"        public static string {dependency.Name} {{ get {{ return {dependency.Name}; }} }}" + Environment.NewLine;
```
does `{dependency.Name}` need to be quoted for it to be a string?  `return "MyName";`

eerhardt replied:
Or is this simply a stackoverflow waiting to happen?

---
In reply to: [107795515](https://github.com/dotnet/sdk/pull/1024#discussion_r107795515) [](ancestors = 107795515)

johnbeisner replied:
This is a reference to the supporting projects - I did get it wrong; fixed now:
**{dependency.Name}.{dependency.Name}Class.Name;**

Commit:
Fixing the library case, project reference.

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,127 @@
+            return ret;
+        }
+
+        bool AllProjectsBuildOnNonWindows(TestProject referencerProject) {
```
nit: brace is on the wrong line.

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,127 @@
+
+        TestProject GetTestProject(string name, string target, bool isSdkProject)
+        {
+            TestProject ret = new TestProject()
```
nit 1: Please call this something more descriptive than `ret`. `testProject` would work fine. 
nit 2: Use `var` when type is obvious

``` C#
var testProject = new TestProject()
```

nguerrera commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToVerifyProjectReferenceCompat.cs:
```diff
@@ -0,0 +1,127 @@
+            if (isSdkProject)
+            {
+                ret.TargetFrameworks = target;
+                //  Workaround for .NET Core 2.0
```
Do we have a bug filed to remove these workarounds once my fix is pulled back in to the repo?
